lib/chain.js

```diff
@@ -2,82 +2,46 @@
 
 module.exports = (api) => {
 
-  const resolver = (resolve, reject) => (
-    (err, result) => {
-      if (err) reject(err);
-      else resolve(result);
-    }
-  );
-
-  class ArrayChain {
-
-    constructor(array) {
-      this._promise = Promise.resolve(array);
-    }
-
-    then(fn) {
-      return this._promise.then(fn);
-    }
-
-    catch(fn) {
-      return this._promise.catch(fn);
-    }
-
-    map(fn) {
-      this._update((array, resolve, reject) => {
-        api.metasync.map(array, fn, resolver(resolve, reject));
-      });
-      return this;
-    }
-
-    filter(fn) {
-      this._update((array, resolve, reject) => {
-        api.metasync.filter(array, fn, resolver(resolve, reject));
-      });
-      return this;
-    }
-
-    reduce(fn, initial) {
-      this._update((array, resolve, reject) => {
-        api.metasync.reduce(array, fn, resolver(resolve, reject), initial);
-      });
-      return this._promise;
-    }
-
-    each(fn) {
-      this._update((array, resolve, reject) => {
-        api.metasync.each(array, fn, resolver(resolve, reject));
-      });
-      return this._promise;
-    }
-
-    series(fn) {
-      this._update((array, resolve, reject) => {
-        api.metasync.series(array, fn, resolver(resolve, reject));
-      });
-      return this._promise;
-    }
-
-    find(fn) {
-      this._update((array, resolve, reject) => {
-        api.metasync.find(array, fn, resolver(resolve, reject));
-      });
-      return this._promise;
-    }
-
-    _update(fn) {
-      this._promise = this._promise.then((array) => (
-        new Promise((resolve, reject) => fn(array, resolve, reject))
-      ));
-    }
-
+  function arrayChain(array, prev = null) {
+
+    let next = null, done = null, fail = null;
+
+    const self = (err, data) => {
+      array = data;
+      if (next) next();
+      if (err) {
+        if (fail) fail(err);
+      } else if (done) {
+        done(data);
+      }
+    };
+
+    if (!prev) process.nextTick(() => self(null, array));
```

Comment:
I also would like to start chain propagation with run() and maybe even without linked list of functors. Here is implementation in branch: [chain-alternative-fire](https://github.com/metarhia/MetaSync/tree/chain-alternative-fire) but it is behind refactored async function so it's so large. See also #49 
We can even create two functions: `run()` for immediate and `compose()` for future use, see #45 for syntax example.